RFC.md

```diff
@@ -0,0 +1,253 @@
+        if use_sigmoid:
+            scores = torch.sigmoid(scores.to(sigmoid_dtype)).to(x.dtype)
+        else:
+            scores = F.softmax(scores.to(softmax_dtype), dim=0).to(x.dtype)
+        # router scores/indices shape [num_experts, tokens_per_expert]
+        top_scores, selected_token_indices = torch.topk(scores, k=self.tokens_per_expert, dim=1)
+
+        # apply the token preprocess function and then run experts forward
+        token_indices = selected_token_indices.reshape(-1, 1).expand(-1, D)
+        # routed input shape [num_experts*tokens_per_expert, hidden_dim]
+        routed_input = torch.gather(x, dim=0, index=token_indices)
+        routed_input = routed_input * top_scores.reshape(-1, 1)
+        return routed_input, token_indices, None,
+```
+
+## Moe Layer
+An MOE layer consists of experts and routers.
+
+**Here's the proposed MoeLayer design in torchtune:**
+```python
+class MoeLayer(nn.Module):
+    def __init__(self, router="token_choice"):
+        self.experts = moe_experts(hidden_dim, model_dim, num_experts=num_experts)
+        self.shared_expert = moe_expert(hidden_dim, model_dim)
+        if router == "token_choice":
+            self.router = TokenChoiceTopKRouter(hidden_dim, num_experts, experts_per_token)
+        elif router == "expert_choice":
+            self.router = ExpertChoiceTopKRouter(hidden_dim, num_experts, tokens_per_expert)
+        else:
+            raise NotImplementedError("This router is not supported yet!")
+
+    def forward(self, x, infernece=False):
+        routed_input, token_indices, num_local_tokens_per_expert = self.router(x)
+        # routed output shape [num_experts*tokens_per_expert, hidden_dim] for EC, [bs*slen*experts_per_token, hidden_dim] for TC
+        routed_output = self.experts(routed_input, num_local_tokens_per_expert=num_local_tokens_per_expert)
+
+        # shared expert
+        if use_shared_expert:
+            out = self.shared_expert(x)
+        else:
+            out = torch.zeros_like(x)
+
+        # add experts output
+        out.data = scatter_add_(
+            out.data,
+            routed_output,
+            selected_indices,
+        )
+        return out
+ ```
+
+## Model builder
+Besides the above components: experts, routers, and MOE layers, we would need a model builder to pull all pieces together to form the Transformer decoder layer and then Transformer decoder:
+
+**Here's the proposed MOE model builder design in torchtune:**
+```python
+def moe(...) -> TransformerDecoder:
+    # Build the decoder associated with the moe model. This includes
+    # - Token embeddings
+    # - num_layers number of TransfomerDecoderLayer block
+    # - RMS Norm layer applied to the ouput of the transfomer
+    # - Final projection into the token space'
+    token_embeddings = nn.Embedding(vocab_size, embed_dim)
+    self_attn = MultiHeadAttention()
+    moe_layer = MoeLayer(router="token_choice") # or MoeLayer(router="expert_choice")
```

Comment:
I would consider an `moe_layer` builder function as well to improve ease of configurability